torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -228,12 +228,16 @@ void Reducer::mark_variable_ready(
     }
   }
 
-  // Queue function to finalize once the final bucket was marked ready.
+  // Run finalizer function once the final bucket was marked ready.
   if (next_bucket_ == buckets_.size()) {
-    // Autograd callbacks can only be registered while the engine is running.
-    AT_ASSERT(called_from_autograd);
-    torch::autograd::Engine::get_default_engine().queue_callback(
-        [=] { this->finalize_backward(); });
+    if (called_from_autograd) {
+      torch::autograd::Engine::get_default_engine().queue_callback([=] {
+        std::lock_guard<std::mutex> lock(this->mutex_);
+        this->finalize_backward();
+      });
+    } else {
+      finalize_backward();
```

Comment:
I am debating with myself whether we should support this or error out with explicit message. What are the use cases for DDP without backward?

Reply:
we should definitely support it. It's possible that for a few iterations there's no backward and for a few iterations there is backward (for example when doing MCTS / rollouts)